ruff_lsp/server.py

```diff
@@ -178,6 +207,149 @@ class VersionModified(NamedTuple):
 ]
 
 
+###
+# Document
+###
+
+
+@enum.unique
+class DocumentKind(enum.Enum):
+    """The kind of document."""
+
+    Text = enum.auto()
+    """A Python file or a cell in a Notebook Document."""
+
+    Notebook = enum.auto()
+    """A Notebook Document."""
+
+
+@dataclass(frozen=True)
+class Document:
+    """A document representing either a Python file, a Notebook cell, or a Notebook."""
+
+    uri: str
+    path: str
+    source: str
+    kind: DocumentKind
+    version: int | None
+
+    @classmethod
+    def from_text_document(cls, text_document: workspace.TextDocument) -> Self:
+        """Create a `Document` from the given Text Document."""
+        return cls(
+            uri=text_document.uri,
+            path=text_document.path,
+            kind=DocumentKind.Text,
+            source=text_document.source,
+            version=text_document.version,
+        )
+
+    @classmethod
+    def from_notebook_document(cls, notebook_document: NotebookDocument) -> Self:
+        """Create a `Document` from the given Notebook Document."""
+        path = uris.to_fs_path(notebook_document.uri)
+        if path is None:
+            # `pygls` raises a `Exception` as well in `workspace.TextDocument`.
+            raise ValueError(
+                f"Unable to convert URI to file path: {notebook_document.uri}"
+            )
+        return cls(
+            uri=notebook_document.uri,
+            path=path,
+            kind=DocumentKind.Notebook,
+            source=_create_notebook_json(notebook_document),
+            version=notebook_document.version,
+        )
+
+    @classmethod
+    def from_uri(cls, uri: str) -> Self:
+        """Create a `Document` from the given URI.
+
+        The URI can be a file URI, a notebook URI, or a cell URI. The function will
+        try to get the notebook document first, and if that fails, it will fallback
+        to the text document.
+        """
+        notebook_document = LSP_SERVER.workspace.get_notebook_document(cell_uri=uri)
+        if notebook_document is None:
+            notebook_document = LSP_SERVER.workspace.get_notebook_document(
+                notebook_uri=uri
+            )
```

Comment:
We can't use:

```python
LSP_SERVER.workspace.get_notebook_document(notebook_uri=uri, cell_uri=uri)
```

Because the method uses _either_ `notebook_uri` or `cell_uri` where `notebook_uri` is preferred over `cell_uri`. While our implementation will first try the `cell_uri` and then the `notebook_uri`.